lib/benchmark/db/mysql.go

```diff
@@ -0,0 +1,148 @@
+	DBUser string
+	// DBName database name where the benchmark queries are going to be
+	// executed.
+	DBName string
+	// InsecureSkipVerify bypasses verification of TLS certificate.
+	InsecureSkipVerify bool
+	// connOptions are the MySQL connection options.
+	connOptions *mysqlConnOptions
+}
+
+func (p *MySQLBenchmark) CheckAndSetDefaults() error {
+	if p.DBService == "" {
+		return trace.BadParameter("database or direct database URI must be provided")
+	}
+
+	if strings.Contains(p.DBService, "://") {
+		var err error
+		p.connOptions, err = parseMySQLConnString(p.DBService)
+		return trace.Wrap(err)
+	}
+
+	if p.DBUser == "" || p.DBName == "" {
+		return trace.BadParameter("must provide and database name and user")
+	}
+
+	return nil
+}
+
+func (p *MySQLBenchmark) BenchBuilder(ctx context.Context, tc *client.TeleportClient) (benchmark.WorkloadFunc, error) {
+	if err := p.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	config, err := p.buildConnectionConfig(ctx, tc)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return func(ctx context.Context) error {
+		conn, err := mysqlclient.Connect(config.host, config.username, config.password, config.database)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		return trace.Wrap(conn.Ping())
+	}, nil
+}
+
+// mysqlConnOptions defines connection configuration for MySQL database.
+type mysqlConnOptions struct {
+	host     string
+	username string
+	password string
+	database string
+}
+
+// buildConnectionConfig generates a connect configuration for the database.
+func (p *MySQLBenchmark) buildConnectionConfig(ctx context.Context, tc *client.TeleportClient) (*mysqlConnOptions, error) {
+	if p.connOptions != nil {
+		return p.connOptions, nil
+	}
+
+	database, err := getDatabase(ctx, tc, p.DBService, types.DatabaseProtocolMySQL)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	dbCert, err := retrieveDatabaseCertificates(ctx, tc, database, p.DBUser, p.DBName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	lp, err := startLocalProxy(ctx, p.InsecureSkipVerify, tc, database.GetProtocol(), dbCert)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &mysqlConnOptions{
+		host:     lp.GetAddr(),
+		username: p.DBUser,
+		database: p.DBName,
+	}, nil
+}
+
+// parseMySQLConnString parses a MySQL URI into connection options.
+// URI Format:
+// [scheme://][user[:[password]]@]host[:port][/schema][?attribute1=value1&attribute2=value2...
+//
+// Reference: https://dev.mysql.com/doc/refman/8.0/en/connecting-using-uri-or-key-value-pairs.html
+func parseMySQLConnString(connString string) (*mysqlConnOptions, error) {
+	parsed, err := url.Parse(connString)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if parsed.Scheme != mysqlProtocolScheme {
+		return nil, trace.BadParameter("scheme %q not supported. Please use %q", parsed.Scheme, mysqlProtocolScheme)
+	}
+
+	password, _ := parsed.User.Password()
```

Comment:
Can we just use `mysql.ParseDSN()` ? Same for postgres we can use `pgconn.ParseConf`

Reply:
It would require making a change to the library to [make it public](https://github.com/gravitational/go-mysql/blob/2dc901e295c0270a472450923ac494edd9065537/driver/driver.go#L48) or start using [`Open`](https://github.com/gravitational/go-mysql/blob/2dc901e295c0270a472450923ac494edd9065537/driver/driver.go#L85) instead of [`Connect`](https://github.com/gravitational/go-mysql/blob/2dc901e295c0270a472450923ac494edd9065537/client/conn.go#L65) (which would trigger a connection before the benchmark start).

Besides that, I'm ok with changing it, but since we're not parsing properties and only supporting username/password authentication parsing the URL seemed the most straightforward approach here.